File: ios/Sources/RynBridgeSpeech/SpeechModule.swift
import Foundation
import RynBridge

public struct SpeechModule: BridgeModule, Sendable {
    public let name = "speech"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init() {
        self.init(provider: DefaultSpeechProvider())
    }

    public init(provider: SpeechProvider) {
        actions = [
            "startRecognition": { payload in
                let language = payload["language"]?.stringValue
                let result = try await provider.startRecognition(language: language)
                return result.toPayload()
            },
            "stopRecognition": { payload in
                guard let sessionId = payload["sessionId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: sessionId")
                }
                let result = try await provider.stopRecognition(sessionId: sessionId)
                return result.toPayload()
            },
            "speak": { payload in
                guard let text = payload["text"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: text")
                }
                let language = payload["language"]?.stringValue
                let rate = payload["rate"]?.doubleValue
                let pitch = payload["pitch"]?.doubleValue
                let voiceId = payload["voiceId"]?.stringValue
                let options = SpeakOptions(text: text, language: language, rate: rate, pitch: pitch, voiceId: voiceId)
                try await provider.speak(options: options)
                return [:]
            },
            "stopSpeaking": { _ in
                provider.stopSpeaking()
                return [:]
            },
            "getVoices": { _ in
                let result = try await provider.getVoices()
                return result.toPayload()
            },
            "requestPermission": { _ in
                let result = try await provider.requestPermission()
                return result.toPayload()
            },
            "getPermissionStatus": { _ in
                let result = try await provider.getPermissionStatus()
                return result.toPayload()
            },
        ]
    }
}

File: android/speech/src/main/kotlin/io/rynbridge/speech/SpeechModule.kt
package io.rynbridge.speech

import android.content.Context
import io.rynbridge.core.*

class SpeechModule(provider: SpeechProvider) : BridgeModule {
    constructor(context: Context) : this(DefaultSpeechProvider(context))

    override val name = "speech"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "startRecognition" to { payload ->
            val language = payload["language"]?.stringValue
            val result = provider.startRecognition(language)
            result.toPayload()
        },
        "stopRecognition" to { payload ->
            val sessionId = payload["sessionId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: sessionId")
            val result = provider.stopRecognition(sessionId)
            result.toPayload()
        },
        "speak" to { payload ->
            val text = payload["text"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: text")
            val language = payload["language"]?.stringValue
            val rate = payload["rate"]?.doubleValue
            val pitch = payload["pitch"]?.doubleValue
            val voiceId = payload["voiceId"]?.stringValue
            val options = SpeakOptions(text = text, language = language, rate = rate, pitch = pitch, voiceId = voiceId)
            provider.speak(options)
            emptyMap()
        },
        "stopSpeaking" to { _ ->
            provider.stopSpeaking()
            emptyMap()
        },
        "getVoices" to { _ ->
            val result = provider.getVoices()
            result.toPayload()
        },
        "requestPermission" to { _ ->
            val result = provider.requestPermission()
            result.toPayload()
        },
        "getPermissionStatus" to { _ ->
            val result = provider.getPermissionStatus()
            result.toPayload()
        }
    )
}
